crates/hotshot/src/traits/networking/push_cdn_network/definition/message_hook.rs

```diff
@@ -0,0 +1,577 @@
+    fn commit_if_necessary(&self, num_required_samples: usize) -> bool {
+        let mut num_samples = self.num_samples.lock();
+
+        // Return early if we don't have enough samples
+        if *num_samples < num_required_samples as u64 {
+            return false;
+        }
+
+        // Calculate the new average
+        let mut num_bytes = self.num_bytes.lock();
+        let new_average = (*num_bytes / *num_samples as f64) as u32;
+
+        // Reset the internal state
+        *num_bytes = 0.0;
+        *num_samples = 0;
+
+        // Cache the new average
+        self.last_committed_average
+            .store(new_average, Ordering::Relaxed);
+
+        // Return the new average
+        true
+    }
+}
+
+/// A sample representing the number of bytes processed per period
+#[derive(Clone)]
+pub struct Sample {
+    /// The number of bytes processed
+    num_bytes: f64,
+
+    /// The time of the last commit
+    last_committed_time: Instant,
+}
+
+impl Sample {
+    /// Create a new `Sample` instance
+    fn new() -> Self {
+        Self {
+            num_bytes: 0.0,
+            last_committed_time: Instant::now(),
+        }
+    }
+
+    /// Add a number of bytes to the sample
+    fn add(&mut self, num: f64) {
+        self.num_bytes += num;
+    }
+
+    /// Commit the sample if necessary, returning the number of bytes processed per commit interval
+    /// and resetting the internal state if we did.
+    fn commit_if_necessary(&mut self, commit_interval: Duration) -> Option<f64> {
+        // Get the elapsed time
+        let elapsed = self.last_committed_time.elapsed();
+
+        // Return early if we aren't past the commit interval
+        if elapsed < commit_interval {
+            return None;
+        }
+
+        // Calculate the new average
+        let value = self.num_bytes / elapsed.div_duration_f64(commit_interval);
+
+        // Reset the internal state
+        self.last_committed_time = Instant::now();
+        self.num_bytes = 0.0;
+
+        // Return the new average
+        Some(value)
+    }
+}
+
+/// The message hook for `HotShot` messages. Each user has a unique
+#[derive(Clone)]
+pub struct HotShotMessageHook {
+    /// The cache for message hashes. We use this to deduplicate a sliding window of
+    /// 100 messages.
+    message_hash_cache: LruCache<u64, ()>,
+
+    /// The reference counter so we can keep track of the number of hooks there are
+    num_hooks: Arc<()>,
+
+    /// The unique identifier for the hook. This is used to uniquely identify the hook
+    /// by the consumer
+    identifier: u64,
+
+    /// The commit interval for a local sample -> global average
+    commit_interval: Duration,
+
+    /// The cache of previously dropped GCRs by identifier. This is used to prevent a user
+    /// from reconnecting and using a brand new GCR instance (empty rate limit)
+    dropped_gcr_cache: Arc<Mutex<LruCache<u64, (Gcr, Gcr)>>>,
```

Comment:
This holds previously-dropped rate limiting states so if a user reconnects they don't just get an entirely new rate limiter